dnf5/commands/check/check.cpp

```diff
@@ -0,0 +1,315 @@
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "check.hpp"
+
+#include <fmt/format.h>
+#include <libdnf5/utils/bgettext/bgettext-lib.h>
+#include <libdnf5/utils/bgettext/bgettext-mark-domain.h>
+#include <rpm/header.h>
+#include <rpm/rpmcli.h>
+#include <rpm/rpmps.h>
+#include <rpm/rpmtag.h>
+#include <rpm/rpmtd.h>
+#include <rpm/rpmte.h>
+#include <rpm/rpmts.h>
+
+#include <iostream>
+#include <map>
+#include <set>
+#include <string>
+#include <vector>
+
+namespace dnf5 {
+
+using namespace libdnf5;
+
+namespace {
+
+enum class ProblemType { MISSING_REQUIRE, CONFLICT, OBSOLETED, DUPLICATE };
+
+// Describes one problem
+struct Problem {
+    ProblemType type;             // type of problem
+    std::string nevra;            // "nevra" causing this problem
+    std::string file_or_provide;  // "file" or "provide" causing this problem
+
+    bool operator<(Problem p) const noexcept {
+        return type < p.type || (type == p.type && (file_or_provide < p.file_or_provide ||
+                                                    (file_or_provide == p.file_or_provide && nevra < p.nevra)));
+    }
+};
+
+// what the program will check
+bool all{true};
+bool dependencies{false};
+bool duplicates{false};
+bool obsoleted{false};
+
+// Contains "provides" to recognize the installonly package.
+// It is initialized with values from the configuration.
+std::vector<std::string> installonly_pkgs_provides;
+
+// Contains a set of problems for each problematic package.
+// The std::map Key is the "nevra" of the problematic package.
+std::map<std::string, std::set<Problem>> problems;
+
+// Contains a list of installed "nevras" for each "name.arch".
+// For "installonly" packages, only one "nevra" is present.
+// Usually contains 1 "nevra" for "name.arch". Multiple "nevras" means "duplicates".
+std::map<std::string, std::vector<std::string>> installed_na_nevras;
+
+
+cli::ArgumentParser::NamedArg * create_option(
+    cli::ArgumentParser & parser, const std::string & name, const std::string & descr, bool & opt_to_modify) {
+    auto option = parser.add_new_named_arg(name);
+    option->set_long_name(name);
+    option->set_description(descr);
+    option->set_parse_hook_func([&opt_to_modify](cli::ArgumentParser::NamedArg *, const char *, const char *) {
+        opt_to_modify = true;
+        return true;
+    });
+    return option;
+}
+
+
+inline bool is_package_installonly(Header h) noexcept {
```

Comment:
I guess that `libsolv` is able to identify installony packages faster. Libsolv uses internally integer instead of string and int comparison is significantly faster.

Reply:
But that means using libsolv. This implementation of the "check" command only uses librpm. The installed packages database is not loaded by libsolv.
Additionally, the `is_package_installonly` function is only called when a duplicate is found. Usually twice on a standard Fedora. When found installed second and third kernel versions.
Two calls of `is_package_installonly` function should be faster than having libsolv load the entire rpmdb and do the string indexing.

Reply:
OK, thank you for explanation. I mark it as reslved